packages/keyring-controller/src/KeyringController.test.ts

```diff
@@ -3538,3 +3565,21 @@ async function withController<ReturnValue>(
     messenger,
   });
 }
+
+/**
+ * Construct a keyring builder with a spy.
+ *
+ * @param KeyringConstructor - The constructor to use for building the keyring.
+ * @returns A keyring builder that uses `jest.fn()` to spy on invocations.
+ */
+function buildKeyringBuilderWithSpy(KeyringConstructor: KeyringClass<Json>): {
+  (): EthKeyring<Json>;
+  type: string;
+} {
+  const keyringBuilderWithSpy: { (): EthKeyring<Json>; type?: string } = jest
+    .fn()
+    .mockImplementation((...args) => new KeyringConstructor(...args));
+  keyringBuilderWithSpy.type = KeyringConstructor.type;
+  // Not sure why TypeScript isn't smart enough to infer that `type` is set here.
+  return keyringBuilderWithSpy as { (): EthKeyring<Json>; type: string };
+}
```

Comment:
Nit:
We can also do:
```
function buildKeyringBuilderWithSpy(KeyringConstructor: KeyringClass<Json>): {
  (): EthKeyring<Json>;
  type: string;
} {
  const keyringBuilderWithSpy = jest
    .fn()
    .mockImplementation(
      (...args) => new KeyringConstructor(...args),
    ) as jest.Mock & { type: string };
  keyringBuilderWithSpy.type = KeyringConstructor.type;
  return keyringBuilderWithSpy;
}

Reply:
Hmm, yes that would also work but it seems equivalent to me (in terms of type safety and readability and whatnot).

I was trying to do it with no assertion.

Reply:
I could do this without an assertion by doing something like

```
keyringBuilderWithSpy.type = KeyringConstructor.type
if (!hasProperty(keyringBuilderWithSpy, 'type')) {
  throw new Error('Missing type');
}
```

TypeScript would be able to infer it then. But that would be very silly to have to do, because we already know that `type` is set by that point (`KeyringConstructor.type` is of type `string`, it's not optional).